python/sglang/srt/layers/attention/trtllm_mla_backend.py

```diff
@@ -681,35 +694,82 @@ def quantize_and_rope_for_fp8(
             dtype=attn_dtype,
         )
 
-        # Key outputs maintain original shapes but with FP8 dtype
-        k_rope_out = k_rope.new_empty(k_rope.shape, dtype=attn_dtype)
-        k_nope_out = k_nope.new_empty(k_nope.shape, dtype=attn_dtype)
-
-        # Apply RoPE and quantize all components in a single fused kernel call
-        # This kernel handles:
-        # 1. RoPE application to q_rope and k_rope using cos_sin_cache and positions
-        # 2. Quantization of all components to FP8 format
-        # 3. Output placement into pre-allocated tensors
-        flashinfer.rope.mla_rope_quantize_fp8(
-            q_rope=q_rope,
-            k_rope=k_rope,
-            q_nope=q_nope,
-            k_nope=k_nope,
-            cos_sin_cache=cos_sin_cache,
-            pos_ids=forward_batch.positions,
-            is_neox=is_neox,
-            quantize_dtype=attn_dtype,
-            # Output tensor slicing: q_out contains [nope_part, rope_part]
-            q_rope_out=q_out[..., self.kv_lora_rank :],  # RoPE part goes to end
-            k_rope_out=k_rope_out,
-            q_nope_out=q_out[..., : self.kv_lora_rank],  # Nope part goes to beginning
-            k_nope_out=k_nope_out,
-            # Quantization scales (set to 1.0 for no additional scaling)
-            quant_scale_q=1.0,
-            quant_scale_kv=1.0,
+        # Determine if we can use the fused kernel (RoPE + quantization + direct KV write)
+        # Conditions: save_kv_cache, layer provided, not NSA packed format, kernel available
+        compute_cap = torch.cuda.get_device_capability()
+        kernel_available = mla_rope_quantize_fp8_fused is not None
+        has_layer = layer is not None
+        nsa_flag = getattr(
+            forward_batch.token_to_kv_pool, "nsa_kv_cache_store_fp8", False
+        )
+
+        # Only enable fusion on SM90+ (H100/B200) due to FP8 hardware support
+        use_fused_kv_write = (
+            kernel_available
+            and (compute_cap[0] >= 9)
```

Comment:
I think we do not need to check it here, since the trtllm_mla backend only supports blackwell?